### PR TITLE
Make ZoomIn and ZoomOut buttons more configurable

### PR DESCRIPTION
**Pull request by annarieger**
Added following options to make zoom buttons config more comfortable:
* `enableZoomInWithBox` / `enableZoomOutWithBox` -> Enables zoom with box without `shift` key
* `enableAnimation` -> Enables animation while zooming
* `dragZoomInteraction` / `dragZoomOutInteraction` -> Custom `ol.interaction.DragZoom` which will be used while zooming with box
* `animationDuration` -> Custom animation duration

Please review @weskamm

### CODE REVIEW

File: src/view/button/ZoomIn.js [hunk 1]
```diff
@@ -69,37 +75,84 @@ Ext.define('BasiGX.view.button.ZoomIn', {
...
-        handler: function() {
-            var me = this;
-            var olMap = me.olMap;
-            var olView;
-            var zoom;
+        /**
+         * When set to true zoom in by clicking and dragging on the map is
+         * enabled. Default is true.
+         */
+        enableZoomInWithBox: true,
```
Comment (marcjansen) [inline src/view/button/ZoomIn.js]: Does this change the current default behaviour?

File: src/view/button/ZoomIn.js [hunk 2]
```diff
@@ -69,37 +75,84 @@ Ext.define('BasiGX.view.button.ZoomIn', {
...
+        /**
+         * Whether zoom action should be animated or not. Default is true.
+         */
+        enableAnimation: true,
```
Comment (marcjansen) [inline src/view/button/ZoomIn.js]: Name suggestion `animate`?

File: src/view/button/ZoomIn.js [hunk 3]
```diff
@@ -69,37 +75,84 @@ Ext.define('BasiGX.view.button.ZoomIn', {
...
+        /**
+         * Reference to ol DragZoom interaction which will be used if
+         * #enableZoomInWithBox is set to true.
+         */
+        dragZoomInteraction: null,
```
Comment (marcjansen) [inline src/view/button/ZoomIn.js]: Do we expect people to pass this, or is it some slot in which we keep our internal reference?

File: src/view/button/ZoomOut.js [hunk 4]
```diff
@@ -73,33 +80,83 @@ Ext.define('BasiGX.view.button.ZoomOut', {
...
+        enableAnimation: true,
+        /**
+         * Reference to ol DragZoom interaction which will be used if
+         * #enableZoomOutWithBox is set to true.
+         */
+        dragZoomOutInteraction: null,
+        /**
+         * Default zoom animation duration in milliseconds.
+         */
+        animationDuration: 500
```
Comment (marcjansen) [inline src/view/button/ZoomOut.js]: Can we put these configs in the parent class?

File: src/view/button/ZoomIn.js (hunk 1 above)
Comment (annarieger) [reply src/view/button/ZoomIn.js]: Currently  zoom is possible by click into map or by dragging a rectangle while `shift` key is pressed, what is, I think, the default behaviour of ol zoom control.

This config doesn't change it, but provides the additional method to zoom by dragging a rectangle without `shift`

File: src/view/button/ZoomIn.js (hunk 3 above)
Comment (annarieger) [reply src/view/button/ZoomIn.js]: If it's not passed, a new `ol.Interaction.DragZoom` will be created and set on button on toggle --> s. [here](https://github.com/terrestris/BasiGX/pull/425/files#diff-cfb68a0905b47011db0dfb25e090a910R108)

**Comment by weskamm** [comment]
I like the changes and agree with the notes of @marcjansen :+1:

**Comment by weskamm** [comment]
As the button now is a togglebutton, we should make sure to keep the old behaviour as default.
Maybe we can keep the old button and just use the togglebutton when the config by user intends it.

**Comment by annarieger** [comment]
> As the button now is a togglebutton, we should make sure to keep the old behaviour as default.
> Maybe we can keep the old button and just use the togglebutton when the config by user intends it.

You're right. I'll take a look at it now.

File: src/view/button/ZoomIn.js (hunk 1 above)
Comment (marcjansen) [reply src/view/button/ZoomIn.js]: But default right now is:

* Click -> zoom in
* click again -> zoom in

Is it now:

* Click -> zoom in, toggle button to pressed
* click again -> untoggle pressed?

I see if I can test this.

File: src/view/button/ZoomIn.js (hunk 2 above)
Comment (annarieger) [reply src/view/button/ZoomIn.js]: Renamed in 030dc92

File: src/view/button/ZoomIn.js (hunk 1 above)
Comment (annarieger) [reply src/view/button/ZoomIn.js]: Original behaviour restored, see also https://github.com/terrestris/BasiGX/pull/425#issuecomment-479861743 from @weskamm and 030dc92

So default behaviour is unchanged now (click on button -> zoom in/out), additional configs like zoom with rectangle without shift are only applicable if button configured as toggle button

File: src/view/button/ZoomOut.js (hunk 4 above)
Comment (annarieger) [reply src/view/button/ZoomOut.js]: Zoom buttons have the `BasiGX.view.button.Base` class as parent which is IMO not suitable for these special configs.
I'm not sure if we need one another extra parent class containing shared configs for only two zoom buttons so I leave it unchanged here.

**Comment by annarieger** [comment]
Original behaviour (zoom button = simple button, each click leads to zoom in/out on the map) is restored. Please kindly review once again, @weskamm & @marcjansen

File: src/view/button/ZoomIn.js [hunk 5]
```diff
@@ -69,37 +80,95 @@ Ext.define('BasiGX.view.button.ZoomIn', {
...
+        afterrender: function() {
+            var me = this;
+            if (Ext.isEmpty(me.olMap)) {
+                me.olMap = BasiGX.util.Map.getMapComponent().getMap();
             }
+        },
+        click: function() {
+            var me = this;
+            // do nothing if configured as toggle button
+            if (!Ext.isEmpty(me.toggleGroup)) {
```
Comment (chrismayer) [inline src/view/button/ZoomIn.js]: I think `toggleGroup` is not the right indicator for this. You can have toggleable button without belonging to a ToggleGroup. [enabeToggle](https://docs.sencha.com/extjs/6.2.0/classic/Ext.button.Button.html#cfg-enableToggle) is the right indicator IMHO.

File: src/view/button/ZoomOut.js [hunk 6]
```diff
@@ -73,33 +85,95 @@ Ext.define('BasiGX.view.button.ZoomOut', {
...
+        afterrender: function () {
+            var me = this;
+            if (Ext.isEmpty(me.olMap)) {
+                me.olMap = BasiGX.util.Map.getMapComponent().getMap();
+            }
+        },
+        click: function () {
+            var me = this;
+            // do nothing if configured as toggle button
+            if (!Ext.isEmpty(me.toggleGroup)) {
```
Comment (chrismayer) [inline src/view/button/ZoomOut.js]: See comment in `ZoomOut` button.

File: src/view/button/ZoomOut.js [hunk 7]
```diff
@@ -73,33 +85,95 @@ Ext.define('BasiGX.view.button.ZoomOut', {
...
+                    me.dragZoomOutInteraction = new ol.interaction.DragZoom({
+                        condition: ol.events.condition.always,
+                        duration: me.animationDuration,
+                        out: true
+                    });
+                    me.olMap.addInteraction(me.dragZoomOutInteraction);
+                }
+            }
+            if (pressed) {
+                me.olMap.on('click', me.zoomOut, me);
```
Comment (chrismayer) [inline src/view/button/ZoomOut.js]: Doesn't this conflict with the `ol.interaction.DragZoom` if `pressed` and `enableZoomOutWithBox` are both set to true?

File: src/view/button/ZoomIn.js [hunk 8]
```diff
@@ -69,37 +80,95 @@ Ext.define('BasiGX.view.button.ZoomIn', {
...
+                if (!me.dragZoomInInteraction) {
+                    me.dragZoomInInteraction = new ol.interaction.DragZoom({
+                        condition: ol.events.condition.always,
+                        duration: me.animationDuration
+                    });
+                    me.olMap.addInteraction(me.dragZoomInInteraction);
+                }
+            }
+            if (pressed) {
+                me.olMap.on('click', me.zoomIn, me);
```
Comment (chrismayer) [inline src/view/button/ZoomIn.js]: Doesn't this conflict with the `ol.interaction.DragZoom` if `pressed` and `enableZoomInWithBox` are both set to true?

File: src/view/button/ZoomIn.js (hunk 5 above)
Comment (annarieger) [reply src/view/button/ZoomIn.js]: Nice catch, thanks! Fixed in 5c11570

File: src/view/button/ZoomOut.js (hunk 6 above)
Comment (annarieger) [reply src/view/button/ZoomOut.js]: s. above

File: src/view/button/ZoomOut.js (hunk 7 above)
Comment (annarieger) [reply src/view/button/ZoomOut.js]: I think it's fine here. Just tested it with a real application - `click` event won't fire if a rectange was drawn per dragging.

Furthermore,  `ol.events.condition.always` which actually always returns `true`, has no effect for single clicks on the map (probably cause bounding box of clicked point is too small to force changing of view resolution -> s. [here](https://github.com/openlayers/openlayers/blob/v4.6.5/src/ol/interaction/dragzoom.js#L74))

File: src/view/button/ZoomIn.js (hunk 8 above)
Comment (annarieger) [reply src/view/button/ZoomIn.js]: S. above

File: src/view/button/ZoomOut.js (hunk 7 above)
Comment (chrismayer) [reply src/view/button/ZoomOut.js]: Thanks for clarification. This is OK for me then.

File: src/view/button/ZoomIn.js (hunk 8 above)
Comment (chrismayer) [reply src/view/button/ZoomIn.js]: Thanks for clarification. This is OK for me then.

**Comment by annarieger** [comment]
Thanks for reviews and remarks, guys. Merging now :pineapple:

**Pull request merged**